### PR TITLE
dev -> master

### PR DESCRIPTION
**Pull request by haakondr**
making PR mainly to highlight the schema changes made

# TODO

- revert back cdom code changes that was removed
- drop cdom metadata that was hardcoded
- delete metadata created by metadata-init
- prefix custom tables with "niva-" ?
- make a script to verify all data from nivabase are in odm2

# Questions

- should we make our foreign keys `on delete cascade`?

Currently it is a bit cumbersome to delete i.e. stations. First we need to delete externalidentifier, then the samplingfeature. If we'd make our foreign keys on delete cascade we only need to delete the external identifier and the samplingfeature gets automatically deleted. Sort of good but could of course be "too easy" to delete

### CODE REVIEW

File: k8s/deployments/deployment-kafka-sink.tmpl.yaml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+#        livenessProbe:
+#          httpGet:
+#            path: /
+#            port: 80
+#          initialDelaySeconds: 0
+#        readinessProbe:
+#          httpGet:
+#            path: /
+#            port: 80
+#          initialDelaySeconds: 0
```
Comment (haakondr) [inline k8s/deployments/deployment-kafka-sink.tmpl.yaml:57]: TODO: setup liveness probe

File: api/src/odm2_postgres_api/queries/core_queries.py [hunk 2]
```diff
@@ -129,31 +135,20 @@ async def find_timeseries(conn: asyncpg.connection, data: LoggersTimeSeriesCreat
     return LoggersTimeSeriesId(**result) if result else None
 
 
-async def find_spectra(conn: asyncpg.connection, samplingfeaturecode: str, raise_if_none=False):
```
Comment (haakondr) [inline api/src/odm2_postgres_api/queries/core_queries.py:132]: This got moved a bit below in the file causing a bit weird diff here. It seems like they are identical from what I can tell

File: api/src/odm2_postgres_api/metadata_init/populate_metadata.py [hunk 3]
```diff
@@ -142,10 +141,4 @@ async def populate_metadata(db_pool):
         for lcv in loggers_controlled_vocabularies():
             await save_controlled_vocab(conn, lcv)
 
-        # cdom
-        for sf in cdom_sampling_features():
-            await save_sampling_features(conn, sf)
-        for dc in cdom_directives():
-            await save_directives(conn, dc)
```
Comment (haakondr) [inline api/src/odm2_postgres_api/metadata_init/populate_metadata.py:149]: these were intentionally removed as the hardcoded metadata was just for testing.

File: api/src/odm2_postgres_api/metadata_init/data/mass_spec/mass_spec_metadata.py [hunk 4]
```diff
@@ -592,20 +528,6 @@ def mass_spec_methods(org_id: int) -> List[MethodsCreate]:
...
-            "methoddescription": "Identifies features previously detected using a fd_* and fdc methods. ",
-            "organizationid": org_id,
-            "annotations": [
-                {
-                    "annotationtypecv": "Method annotation",
-                    "annotationtext": "The json field holds the parameters with which this method will be executed",
-                    "annotationjson": json.dumps({"id_feature_wgts": [1, 1, 1, 1, 1, 1, 1], "ulsa": "6cbad51"}),
-                }
-            ],
-        },
```
Comment (haakondr) [inline api/src/odm2_postgres_api/metadata_init/data/mass_spec/mass_spec_metadata.py:608]: this should probably not have been removed. @ZofiaRudjord ? also got removed in commit https://github.com/NIVANorge/odm2-postgres-api/commit/925941d731627b290e47853aef735398f118d1ea#diff-09a349c0efe69c877b2f62c0ee08e62a0df0b367b5cb7b6a7a876b9f75f581faL258

File: api/src/odm2_postgres_api/metadata_init/data/general/variables.py [hunk 5]
```diff
@@ -74,13 +74,6 @@
         "variablecode": "010",
         "nodatavalue": -9999,
     },
-    {
-        "variabletypecv": "Chemistry",
-        "variablenamecv": "Absorbance",
-        "variabledefinition": "This variable mesaures absorption coefficient in a water sample",
-        "variablecode": "Absorbance",
-        "nodatavalue": -9999,
-    },
```
Comment (haakondr) [inline api/src/odm2_postgres_api/metadata_init/data/general/variables.py:83]: should not be deleted?

File: api/src/odm2_postgres_api/metadata_init/data/general/units.py [hunk 6]
```diff
@@ -116,11 +116,6 @@
         "unitsabbreviation": "m",
         "unitsname": "meter",
     },
-    {
-        "unitstypecv": "Length",
-        "unitsabbreviation": "nm",
-        "unitsname": "nanometer",
-    },
```
Comment (haakondr) [inline api/src/odm2_postgres_api/metadata_init/data/general/units.py:123]: should not be deleted?

File: api/src/odm2_postgres_api/metadata_init/data/general/variables.py (hunk 5 above)
Comment (ZofiaRudjord) [reply api/src/odm2_postgres_api/metadata_init/data/general/variables.py:83]: That should be in

File: api/src/odm2_postgres_api/metadata_init/data/mass_spec/mass_spec_metadata.py (hunk 4 above)
Comment (ZofiaRudjord) [reply api/src/odm2_postgres_api/metadata_init/data/mass_spec/mass_spec_metadata.py:608]: @jemmima have you deleted it intensionally? (if so why?)

File: api/src/odm2_postgres_api/metadata_init/data/general/units.py (hunk 6 above)
Comment (ZofiaRudjord) [reply api/src/odm2_postgres_api/metadata_init/data/general/units.py:123]: not deleted

File: api/src/odm2_postgres_api/metadata_init/data/mass_spec/mass_spec_metadata.py [hunk 7]
```diff
@@ -255,16 +255,6 @@ def mass_spec_methods(org_id: int) -> List[MethodsCreate]:
...
-            "methodtypecv": "Derivation",
-            "methodcode": "mass_spec:combine_features",
-            "methodname": "feature combination",
-            "methoddescription": "After feature deconvolution has completed for multiple replicas, "
-            "a list of peaks is output for each replica. These are combined into a single "
-            "list of peaks, in which non-replicating peaks are removed and replicating peaks "
-            "are averaged. This is the input peak list for feature identification.",
-            "organizationid": org_id,
-        },
         {
```
Comment (ZofiaRudjord) [inline api/src/odm2_postgres_api/metadata_init/data/mass_spec/mass_spec_metadata.py:258]: This also looks like deleted by accident

File: api/src/odm2_postgres_api/metadata_init/data/mass_spec/mass_spec_metadata.py [hunk 8]
```diff
@@ -350,35 +340,6 @@ def mass_spec_methods(org_id: int) -> List[MethodsCreate]:
...
-                            "int_var": 5,
-                            "s2n": 2,
-                            "min_nscan": 3,
-                            "safd": "8cb7c42",
-                        }
-                    ),
-                }
-            ],
-        },
         {
```
Comment (ZofiaRudjord) [inline api/src/odm2_postgres_api/metadata_init/data/mass_spec/mass_spec_metadata.py:343]: should not have been deleted

File: api/src/odm2_postgres_api/metadata_init/data/mass_spec/mass_spec_metadata.py [hunk 9]
```diff
@@ -525,31 +486,6 @@ def mass_spec_methods(org_id: int) -> List[MethodsCreate]:
...
-                            "ret_win_per": 0.5,
-                            "delta_mass": 0.004,
-                            "min_int_frag": 300,
-                            "r_thresh": 0.75,
-                            "compcreate": "9b94edf",
-                        }
-                    ),
-                }
-            ],
-        },
```
Comment (ZofiaRudjord) [inline api/src/odm2_postgres_api/metadata_init/data/mass_spec/mass_spec_metadata.py:552]: should not have been deleted

**Review by ZofiaRudjord — CHANGES_REQUESTED** [review]
I looked through the PR and commented on the lines which were accidentally deleted and should be brought back.
There were several instances which showed differences caused by formatting. It is easy to overlook if sth was deleted for those cases.

File: api/src/odm2_postgres_api/metadata_init/data/mass_spec/mass_spec_metadata.py (hunk 4 above)
Comment (jemmima) [reply api/src/odm2_postgres_api/metadata_init/data/mass_spec/mass_spec_metadata.py:608]: I've been trying to work out how this happened - I think it must have been a rebase error as I didn't remove any of these (intentionally), and I'm surprised to see them included in that commit, but it was the earliest commit included it the rebase so that is currently my best guess as the root cause.

**Review by jemmima — CHANGES_REQUESTED** [review]
I have odm2-postgres-api working now after reverting the directives table changes (for now) but I am having trouble running the tests. I have never got them working but we should validate they still work (ideally someone could help me :D)

File: api/src/odm2_postgres_api/metadata_init/data/mass_spec/mass_spec_metadata.py (hunk 4 above)
Comment (haakondr) [reply api/src/odm2_postgres_api/metadata_init/data/mass_spec/mass_spec_metadata.py:608]: merge errors happen. In hindsight we should have done a better job to make sure our branch was "clean" before we started. From my experience things like this always happen when working with long-lived branches. 

The important thing is that we have a good overview before deploying to production.

File: api/src/odm2_postgres_api/db_init/ODM2_for_PostgreSQL.sql [hunk 10]
```diff
@@ -2367,8 +2367,84 @@ alter table ODM2.RelatedTaxonomicClassifiers add constraint fk_RelatedTaxonomicC
 foreign key (RelationshipTypeCV) References ODM2.CV_RelationshipType (Name)
 on update no Action on delete RESTRICT;
 
-alter table ODM2.Directives add column directivename varchar (255) UNIQUE;
```
[the remark this answers is not in the document]
Comment (jemmima) [reply api/src/odm2_postgres_api/db_init/ODM2_for_PostgreSQL.sql:2370]: Done in [bdee0ad](https://github.com/NIVANorge/odm2-postgres-api/pull/206/commits/bdee0ad702cf75e9002183186208e1d4f9c4f378)

File: api/src/odm2_postgres_api/schemas/schemas.py [hunk 11]
```diff
@@ -333,10 +333,11 @@ def check_utc_offset(cls, enddatetime: dt.datetime, values):
 class SamplingFeaturesCreate(BaseModel):
     samplingfeatureuuid: uuid.UUID
     samplingfeaturetypecv: constr(max_length=255)  # type: ignore
-    samplingfeaturecode: constr(max_length=100)  # type: ignore
+    samplingfeaturecode: constr(max_length=50)  # type: ignore
     samplingfeaturename: constr(max_length=255) = None  # type: ignore
     samplingfeaturedescription: constr(max_length=5000) = None  # type: ignore
     samplingfeaturegeotypecv: constr(max_length=255) = None  # type: ignore
+    featuregeometry: constr(max_length=255) = None  # type: ignore
```
[the remark this answers is not in the document]
Comment (haakondr) [reply api/src/odm2_postgres_api/schemas/schemas.py]: I removed this line in https://github.com/NIVANorge/odm2-postgres-api/pull/206/commits/d74f5ddb317215afe910ccc1be8cade5d156764f to make the tests pass. We should probably add inn location info at some time, but I suggest we sit down and discuss the API structure first

[the remark this answers is not in the document]
Comment (jemmima) [reply api/src/odm2_postgres_api/schemas/schemas.py]: I'm actually having trouble just running the tests locally - any ideas?: 

```test setup failed
self = <pytest_docker_compose.DockerComposePlugin object at 0x7fafd8627a60>
request = <SubRequest 'docker_project' for <Function test_post_new_indices>>

    @pytest.fixture(scope="session")
    def docker_project(self, request):
        """
        Builds the Docker project if necessary, once per session.
    
        Returns the project instance, which can be used to start and stop
        the Docker containers.
        """
    
        compose_files = []
    
        for docker_compose in [Path(f) for f in request.config.getoption("docker_compose").split(',')]:
            if docker_compose.is_dir():
                docker_compose /= "docker-compose.yml"
    
            if not docker_compose.is_file():
>               raise ValueError(
                    "Unable to find `{docker_compose}` "
                    "for integration tests.".format(
                        docker_compose=docker_compose.absolute(),
                    ),
                )
E               ValueError: Unable to find `/Users/jemmima/odm2/odm2-postgres-api/api/tests/docker-compose.yml` for integration tests.```

**Review by jemmima — COMMENTED** [review]
Getting the following error on a fresh environment:

```Attaching to odm2-postgres-api_odm2_postgres_api_1
odm2_postgres_api_1  | Traceback (most recent call last):
odm2_postgres_api_1  |   File "/app/src/odm2_postgres_api/main.py", line 21, in <module>
odm2_postgres_api_1  |     from odm2_postgres_api.app import app
odm2_postgres_api_1  | ModuleNotFoundError: No module named 'odm2_postgres_api'
```
I'm actually a bit stuck and confused about why it's not working locally for me but it is working in ci?

**Comment by jemmima** [comment]
> Member

The image was not coming up locally or in nivatest since pip install command was removed from /api Dockerfile in dc23c9a72e4d9b7119d17aeb25560951623c5dc3. This meant the api code was not added to the module search path and api container was failing to come up. Fixed in [Fix local docker env](https://github.com/NIVANorge/odm2-postgres-api/pull/206/commits/89bdb19377251b692e79c9fba22632f79c5468c1)

File: api/src/odm2_postgres_api/schemas/schemas.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (haakondr) [reply api/src/odm2_postgres_api/schemas/schemas.py]: This has to do with your current working directory. You need to run the tests from api/ folder (and not tests folder).

The plugin used in odm2 API called `pytest-docker-compose` is not very flexible when it comes to defining the location of the docker-compose.yml file. This is the main reason I've not used this for kafka-sink (and observations-api)

**Comment by haakondr** [comment]
> > Getting the following error on a fresh environment:
> > ```
> > odm2_postgres_api_1  | Traceback (most recent call last):
> > odm2_postgres_api_1  |   File "/app/src/odm2_postgres_api/main.py", line 21, in <module>
> > odm2_postgres_api_1  |     from odm2_postgres_api.app import app
> > odm2_postgres_api_1  | ModuleNotFoundError: No module named 'odm2_postgres_api'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'm actually a bit stuck and confused about why it's not working locally for me but it is working in ci?
> 
> The image was not coming up locally or in nivatest since pip install command was removed from /api Dockerfile in [dc23c9a](https://github.com/NIVANorge/odm2-postgres-api/commit/dc23c9a72e4d9b7119d17aeb25560951623c5dc3). This meant the api code was not added to the module search path and api container was failing to come up. Fixed in [89bdb19](https://github.com/NIVANorge/odm2-postgres-api/commit/89bdb19377251b692e79c9fba22632f79c5468c1)

Ah, sorry about that. Not sure why we're doing this though. I must admit I've never fully understood how to handle relative paths for imports in python. 

The problem here is how do we define which file the following import points to:

```
from odm2_postgres_api.app import app
```

The way we've solved this in kafka-sink and observations-api is to move the entrypoint (main.py) one directory up, so that relative to main.py's directory python can look for the folder `odm2_postgres_api` and the file app.py.

The alternative solution which is done here is to do a pip install on the odm2_postgres_api folder and make all the files available as a module on the local python installation. This works and also allow us to put app.py in whatever location we want. 

One drawback to this is that we need to run `pip install .` for all code changes, or alternatively run `pip install -e .` which will cause changes to apply. This is not very difficult to do, but I've found the more commands you need to run, the higher the chance something gets wrong. Setting up virtual environments and handling that is enough trouble I think.

TLDR; i am not sure about what the best way to do this, but we should probably land on one way to do it.

**Comment by haakondr** [comment]
I added back in code accidentally deleted in 7f1090f, hopefully everything should be ok now

**Comment by jemmima** [comment]
@haakondr  did you mean to commit these secrets? https://github.com/NIVANorge/odm2-postgres-api/tree/dev/api/src/odm2_postgres_api/config

**Comment by haakondr** [comment]
> @haakondr did you mean to commit these secrets? https://github.com/NIVANorge/odm2-postgres-api/tree/dev/api/src/odm2_postgres_api/config

These are used for local development so should not be a problem

File: api/src/odm2_postgres_api/metadata_init/data/mass_spec/mass_spec_metadata.py (hunk 8 above)
Comment (jemmima) [reply api/src/odm2_postgres_api/metadata_init/data/mass_spec/mass_spec_metadata.py:358]: I deleted this as we don't want not working algo versions

**Review by jemmima — APPROVED** [review]

**Comment by jemmima** [comment]
> I looked through the PR and commented on the lines which were accidentally deleted and should be brought back. There were several instances which showed differences caused by formatting. It is easy to overlook if sth was deleted for those cases.

I have compared with how dev looked like before the kafka branch was merged, and I think everything is there. If something is missed it is not difficult to put back, but I have spent an hour going through the diffs and I feel confident enough given the risk is low.
